etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R

```diff
@@ -173,28 +226,162 @@ process_parcel_file <- function(s3_bucket_uri,
         filter(!is.na(lon) & !is.na(x_3435)) %>%
         bind_rows(spatial_df_missing)
     }
+    tictoc::toc()
+
+    tictoc::tic(paste("Calculated shape features:", file_year))
+    # Simplify the planar geometry to reduce noise in shape-based features
+    # calculated below. Convert to a data.table of the vertices of each polygon.
+    # Using data.table here because it's much faster than dplyr
+    spatial_mat_coords <- spatial_df_merged %>%
+      st_set_geometry("geometry_3435") %>%
+      st_simplify(dTolerance = 2, preserveTopology = TRUE) %>%
+      st_cast("MULTIPOLYGON") %>%
+      st_coordinates() %>%
+      as.data.table()
+
+    # Use the coordinate vertices to calculate additional features about
+    # the shape of the polygon
+    spatial_mat_calc <- spatial_mat_coords[
+      ,
+      # Calculate the interior angle between each pair of points
+      angle := calculate_angles(as.matrix(.SD)),
+      .SDcols = c("X", "Y"),
+      by = c("L1", "L2", "L3")
+    ][
+      ,
+      # Some vertex angles are close to 180 or 360 degrees, which are likely to
+      # be artifacts of the data and not true vertices. We don't want to count
+      # these straight angles in our features, BUT we also can't drop all of
+      # them since some polygons are basically just lines. So, we drop any
+      # straight angle vertices from any polygon with >= 3 non-straight angles
+      # (which is the minimum number of vertices needed to make a polygon)
+      straight_angle := (
+        (
+          data.table::between(angle, 0, 1) |
+            data.table::between(angle, 359, 360)
+        ) | data.table::between(angle, 179, 181)
+      )
+    ][
+      ,
+      num_non_straight_angle := sum(!straight_angle, na.rm = TRUE),
+      by = c("L1", "L2", "L3")
+    ][
+      !straight_angle | num_non_straight_angle < 3 | is.na(angle),
```

Comment:
Per @jeancochrane's prior comment, this calculates the interior vertex angle and removes straight angles _before_ calculating the rest of the features (edge length, distance to centroid, etc.). This yields way more accurate feature results, nearly all the PINs with a high angle SD are wacky things like this:

![f71be41a-18b1-4129-b980-a84ec9997a92](https://github.com/ccao-data/data-architecture/assets/31494343/4e39dedb-a674-4a6c-9b4e-cc682dc8a14a)